# Orthologues
**Adrián Barreno Sánchez (adrian.barreno@alumnos.upm.es), Pablo Mata Aroco (p.mata@alumnos.upm.es), Alberto González Delgado (alberto.gondelgado@alumnos.upm.es), Julian Elijah Politsch (julian.politsch@alumnos.upm.es), Angelo D'Angelo (angelo.dangelo@alumnos.upm.es)**


## What is Orthologues?

Orthologues is a computer program designed in ruby to 

## How to run Orthologues?

Download the code as [README file for assignment answers](../README.md) indicates.

## BLAST options and filtering criteria

After a quick search of the scientific literature, we came up with a comparative analysis of BLAST options for the detection of orthologous genes using the Reciprocal Best Hit approach. [Gabriel Moreno-Hagelsieb and Kristen Latimer (2009)](https://academic.oup.com/bioinformatics/article/24/3/319/252715?login=false) tuned two BLAST parameters that significantly affected the alignment scores and, therefore, the election of best hits: 1. the filtering of low information sequence segments and the algorithm used to produce the alignment. Based on the results obtained, the authors recommended using a combination of soft filtering with a Smith–Waterman final alignment algorithm (-F "m S" -s T) options, which provides the highest number of orthologs and the minimal error rates.

## Requirements:

* [BioRuby-Gem](https://rubygems.org/gems/bio-gem/versions/1.3.6)
```
gem install bio-gem
```
* [CSV-gem](https://rubygems.org/gems/csv?locale=es)

```
gem install csv
```
* [Fasta files](https://drive.google.com/drive/folders/0B7FLMiAz5IXPTWJDSkk1MTFPMjg?resourcekey=0-yhXCH6PxXIvg9xwMSolpMw)

* Create the BLAST-formatted databases for BLAST analysis. We need to create a database for each reference proteome using the following command:
```
makeblastdb -in ./files/{Fasta File} -dbtype 'prot' -out ./databases/{BLAST Database}
```

## Usage:

To run the program, execute the following command (inside orthologues/ folder), adding the arguments recquired:

```
cd assignments_answers/orthologues/
```
```
ruby main.rb <Arabidopsis.fa>  <S_pombe.fa> <output_file> 
```
**Arguments:**
1. **Arabidopsis_Database**
2. **S_pombe_Database**
3. **Arabidopsis.fa** 
4. **S_pombe.fa**
2. **Output_file**: file where the report will be saved. Here there is an  [ouput txt file](documents/) that shows the file generated by the program.

To run the program using the files contained in [files folder](documents/), just execute the following command:
```
ruby main.rb documents/ArabidopsisSubNetwork_GeneList.txt ./documents/output.txt
```
If output pathway specified already exists, the program will ask you if you want to overwrite it. Y/y (yes) or N/n (no) input is expected.


## Output
The output is verbose, it will be saved into a [output txt file](documents/). 

## References

* Gabriel Moreno-Hagelsieb, Kristen Latimer, Choosing BLAST options for better detection of orthologs as reciprocal best hits, Bioinformatics, Volume 24, Issue 3, 1 February 2008, Pages 319–324, https://doi.org/10.1093/bioinformatics/btm585

